### DISCUSSION TITLE
期望证书手动创建时选择本地文件可自动化更新

### DISCUSSION THREAD

**Discussion by jiuyue52** [1Panel-dev/1Panel#6299] [Q&A]
### 1Panel 版本

专业版 v1.10.14-lts

### 请描述您的需求或者改进建议

由于1panel暂无精力适配华为云dns账号，因此使用certd程序自动化申请证书，可以推送到服务器本地

**1panel这边也能选择本地文件，但是当本地证书文件更新，1panel要想更新还需要再次选择证书路径才能获取最新证书信息，不能自动获取本地新的证书文件**
![image](https://github.com/user-attachments/assets/a50f6b97-ea88-44b0-9eaa-b0e1025d6553)

需求&改进建议：
**在选择手动上传且选择为本地文件时，此更新按钮点击后可立即获取新的本地证书文件**
![image](https://github.com/user-attachments/assets/795fec7f-bcbc-4cf2-a239-d27eaafe9867)
**在证书详情界面显示本地证书路径，以及可以配置 自动化任务：定时自动获取本地证书文件，以达到自动更新**
![image](https://github.com/user-attachments/assets/abfafb42-40a7-4dec-9e18-974926d97e8f)

### 请描述你建议的实现方案

**在选择手动上传且选择为服务器文件时，更新按钮点击后可立即获取新的本地证书文件**
**在证书详情界面显示本地证书路径，以及可以配置 自动化任务：定时自动获取本地证书文件，以达到自动更新**

**Comment by zhengkunwang223** [comment]
你用了证书自动申请证书 可以直接推送到网站证书目录吧 这样就省去了这些中间步骤

**Reply by alexallen1** [reply]
我看了你回答‘你用了证书自动申请证书 可以直接推送到网站证书目录吧 ’

我发现这样还是不行：
我把fullchain.pem和privkey.pem推送到sites/www123.cn/ssl
https-导入方式-选择ssl文件夹里面2个证书。
更新ssl里面证书文件后。1panel看到的证书日期并未更新。

请问正确做法应该是怎样，才能让1panel读取会更新的证书文件

**Comment by jiuyue52** [comment]
> 你用了证书自动申请证书 可以直接推送到网站证书目录吧 这样就省去了这些中间步骤

有理，也不是不行hhh（ 但这样网站多的话，要添加好多推送任务，想着推送一遍一劳永逸，后边新增网站也不用重新添加推送任务😂😂

**Comment by ColinZeb** [comment]
附议，这个功能很需要。目前1panel也没有设计接口和apikey无法手动实现

**Reply by zhengkunwang223** [reply]
> 附议，这个功能很需要。目前1panel也没有设计接口和apikey无法手动实现

rest-api 文档地址  <1Panel访问地址>/1panel/swagger/index.html

**Comment by ColinZeb** [comment]
我实现了一个，方便后面搜到这的人用，我分享一下：
``` shell
#!/bin/bash

# 定义用户名和密码
USERNAME="****"  
PASSWORD="****"  

# 安全入口 EntranceCode（Base64 ）
ENTRANCE_CODE="****"  

# 1Panel API 地址
LOGIN_URL="https://****.cn/api/v1/auth/login"  
UPLOAD_URL="https://****.cn/api/v1/websites/ssl/upload"  

# 定义证书相关信息
PRIVATE_KEY_PATH="/etc/cert/****/cert.key"  
CERTIFICATE_PATH="/etc/cert/****/cert.crt"  
SSL_ID=1
DESCRIPTION="sync from certd:tx4"

# 构建登录请求的 JSON 数据
LOGIN_DATA=$(cat <<EOF
{
    "name": "$USERNAME",
    "password": "$PASSWORD",
    "language":"zh",
    "authMethod": "jwt"
}
EOF
)
# 使用 curl 发送登录请求
RESPONSE=$(curl -s -X POST "$LOGIN_URL" \
    -H "Content-Type: application/json" \
    -H "EntranceCode: $ENTRANCE_CODE" \
    -d "$LOGIN_DATA")
# 解析返回的 token
TOKEN=$(echo $RESPONSE | grep -Po '(?<="token":")[^"]*')

# 检查是否登录成功
if [ -n "$TOKEN" ]; then
    echo "登录成功，token: $TOKEN"

    # 读取私钥和证书内容
#    PRIVATE_KEY_CONTENT=$(cat "$PRIVATE_KEY_PATH")
#    CERTIFICATE_CONTENT=$(cat "$CERTIFICATE_PATH")

    # 构建上传证书的请求数据
    UPLOAD_DATA=$(cat <<EOF
{
    "privateKey": "",
    "certificate": "",
    "privateKeyPath": "$PRIVATE_KEY_PATH",
    "certificatePath": "$CERTIFICATE_PATH",
    "type": "local",
    "sslID": $SSL_ID,
    "description": "$DESCRIPTION"
}
EOF
    )

    # 发送证书上传请求
    UPLOAD_RESPONSE=$(curl -s -X POST "$UPLOAD_URL" \
        -H "PanelAuthorization: $TOKEN" \
        -H "Content-Type: application/json" \
        -d "$UPLOAD_DATA")

    # 输出上传结果
    echo "证书上传响应: $UPLOAD_RESPONSE"
else
    echo "登录失败，请检查用户名或密码"
    echo "失败响应: $RESPONSE"
fi

```

**Reply by dextercai** [reply]
如果使用远程上传的方式的，应该把这部分
```
PRIVATE_KEY_CONTENT=$(cat "$PRIVATE_KEY_PATH")
CERTIFICATE_CONTENT=$(cat "$CERTIFICATE_PATH")
```
修改成
```
PRIVATE_KEY_CONTENT=$(cat "$PRIVATE_KEY_PATH" | awk '{printf "%s\\n", $0}' | sed 's/\\n$//')
CERTIFICATE_CONTENT=$(cat "$CERTIFICATE_PATH" | awk '{printf "%s\\n", $0}' | sed 's/\\n$//')
```

以避免JSON的转义问题

**Reply by pa1318** [reply]
> 我实现了一个，方便后面搜到这的人用，我分享一下：
> 
> ```shell
> #!/bin/bash
> 
> # 定义用户名和密码
> USERNAME="****"  
> PASSWORD="****"  
> 
> # 安全入口 EntranceCode（Base64 ）
> ENTRANCE_CODE="****"  
> 
> # 1Panel API 地址
> LOGIN_URL="https://****.cn/api/v1/auth/login"  
> UPLOAD_URL="https://****.cn/api/v1/websites/ssl/upload"  
> 
> # 定义证书相关信息
> PRIVATE_KEY_PATH="/etc/cert/****/cert.key"  
> CERTIFICATE_PATH="/etc/cert/****/cert.crt"  
> SSL_ID=1
> DESCRIPTION="sync from certd:tx4"
> 
> # 构建登录请求的 JSON 数据
> LOGIN_DATA=$(cat <<EOF
> {
>     "name": "$USERNAME",
>     "password": "$PASSWORD",
>     "language":"zh",
>     "authMethod": "jwt"
> }
> EOF
> )
> # 使用 curl 发送登录请求
> RESPONSE=$(curl -s -X POST "$LOGIN_URL" \
>     -H "Content-Type: application/json" \
>     -H "EntranceCode: $ENTRANCE_CODE" \
>     -d "$LOGIN_DATA")
> # 解析返回的 token
> TOKEN=$(echo $RESPONSE | grep -Po '(?<="token":")[^"]*')
> 
> # 检查是否登录成功
> if [ -n "$TOKEN" ]; then
>     echo "登录成功，token: $TOKEN"
> 
>     # 读取私钥和证书内容
> #    PRIVATE_KEY_CONTENT=$(cat "$PRIVATE_KEY_PATH")
> #    CERTIFICATE_CONTENT=$(cat "$CERTIFICATE_PATH")
> 
>     # 构建上传证书的请求数据
>     UPLOAD_DATA=$(cat <<EOF
> {
>     "privateKey": "",
>     "certificate": "",
>     "privateKeyPath": "$PRIVATE_KEY_PATH",
>     "certificatePath": "$CERTIFICATE_PATH",
>     "type": "local",
>     "sslID": $SSL_ID,
>     "description": "$DESCRIPTION"
> }
> EOF
>     )
> 
>     # 发送证书上传请求
>     UPLOAD_RESPONSE=$(curl -s -X POST "$UPLOAD_URL" \
>         -H "PanelAuthorization: $TOKEN" \
>         -H "Content-Type: application/json" \
>         -d "$UPLOAD_DATA")
> 
>     # 输出上传结果
>     echo "证书上传响应: $UPLOAD_RESPONSE"
> else
>     echo "登录失败，请检查用户名或密码"
>     echo "失败响应: $RESPONSE"
> fi
> ```

被我翻到这个回复，我跟楼主有同样的需求，但是你这个利用api实现起来似乎有点复杂，不知道有没有更好的方案。

**Comment by alexallen1** [comment]
同问，我这边可以一直推送证书文件到1panel的vps本地。但1panel怎么实现读取本地证书文件，搞了好久也搞不通

**Reply by pa1318** [reply]
> 同问，我这边可以一直推送证书文件到1panel的vps本地。但1panel怎么实现读取本地证书文件，搞了好久也搞不通

我也是有多台服务器同步分发证书，不想每台都手动去上传一下更新，1panel的证书似乎是写到数据库里的，不能直接读取本地证书文件。
